content/doc/developer/security/form-validation.adoc

```diff
@@ -0,0 +1,109 @@
+---
+title: Securely implementing form validation
+layout: developerguide
+---
+
+This guide looks at form validation from a security point of view:
+What are the considerations for securing form validation, and how can they be best implemented?
+
+First, let's look at a typical form field defined in Jelly, and its validation method in the object's descriptor.
+This forms the basis for the rest of this guide.
+
+[source, xml]
+----
+<f:field name="foo">
+  <f:textbox />
+</f:field>
+----
+
+[source, java]
+----
+public FormValidation doCheckFoo(@QueryParameter String value) {
+  if (Util.fixEmptyAndTrim(value) == null) {
+    return FormValidation.error("foo cannot be empty");
+  }
+  return FormValidation.ok();
+}
+----
+
+Stapler routing and Jenkins behavior will result in this method being routed at a URL like `/descriptorByName/f.q.ClassName/checkFoo`.
+This URL will contain a minimal HTML snippet with a form validation message (if present) based on the presence and value of the `value` query parameter.
+
+=== Potential Security Issues
+
+==== Returning Private Information
+
+By default, any user with Overall/Read access can access form validation methods like the example method above.
+Some form validation methods could be used to infer private information, and as such may need to be protected from unauthorized access.
+
+If a form validation method is supposed to only be accessible to users with specific permissions, a permission check should be added.
+
+==== Side Effects
```

Comment:
I would also explicitly mention that the requests may be invoked from outside Jenkins, and hence it's the developer's responsibility to do proper parameter filtering before they do something dangerous or something exposing/using sensitive data in the handler code

Reply:
So my proposal would be to just rename the section to "Parameter handling risks" or so